scio-smb/src/main/scala/com/spotify/scio/smb/syntax/SortMergeBucketSCollectionSyntax.scala

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2019 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.smb.syntax
+
+import com.spotify.scio.coders.Coder
+import com.spotify.scio.io.ClosedTap
+import com.spotify.scio.testing.TestDataManager
+import com.spotify.scio.util.ScioUtil
+import com.spotify.scio.values._
+import org.apache.avro.generic.GenericRecord
+import org.apache.avro.specific.SpecificRecordBase
+import org.apache.beam.sdk.extensions.smb.BucketMetadata.HashType
+import org.apache.beam.sdk.extensions.smb.{AvroSortedBucketIO, SortedBucketSink}
+
+import scala.reflect.ClassTag
+
+trait SortMergeBucketSCollectionSyntax {
+  implicit def toGenericAvroCollection(
+    data: SCollection[GenericRecord]
+  ): SortedBucketGenericAvroSCollection =
+    new SortedBucketGenericAvroSCollection(data.asInstanceOf[SCollection[GenericRecord]])
+
+  implicit def toSpecificAvroCollection[T <: SpecificRecordBase: ClassTag: Coder](
+    data: SCollection[T]
+  ): SortedBucketSpecificAvroSCollection[T] =
+    new SortedBucketSpecificAvroSCollection[T](data)
+}
+
+private[smb] object SortMergeBucketSCollectionSyntax {
+  // Adapted from ScioUtil
+  def toPathFragment(path: String): String =
+    if (path.endsWith("/")) s"${path}bucket-*-shard-*.avro"
+    else s"$path/bucket-*-shard-*.avro"
+}
+
+final class SortedBucketGenericAvroSCollection(private val self: SCollection[GenericRecord]) {
+  import com.spotify.scio.avro.{GenericRecordIO, GenericRecordTap}
+  import com.spotify.scio.smb.syntax.SortMergeBucketSCollectionSyntax._
+  import org.apache.avro.Schema
+  import org.apache.avro.file.CodecFactory
+
+  def saveAsSortedBucket[K: Coder](
```

Comment:
We'll probably have to rename these methods so they don't cause ambiguous overloading issues.
Either that, or maybe we can have a type class to encapsulate `BucketMetadata`, `FileOps`, etc.?

Something like:
```
trait SmbIO[T] {
  def bucketMetdata: BucketMetadata[T]
  def fileOps: FileOperations[T]
}

object SmbIO {
  def forJson(keyField: String): SmbIO[String]
  def forGenericRecord(...)
  ...
}
```

`schema`, `codec` (specific to Avro, Parquet), `CompressionType` (specific to TFRecord, JSON), `keyField` (might be `List[String]` in the case of feature store?) can be IO specific, while `hashType`, `numBuckets`, `numShards` are common to all IOs right?

Reply:
On second though, can this type class encapsulate `SortedBucketIO.{Read,Write}` both?

```
trait SmbIO[T] {
  def read(params ...): SortedBucketIO.Read[T]
  def write(params ...): SortedBucketIO.Write[T]
}
```

Reply:
typeclass could work--we could also do something simpler and just put the Avro/Json/TF sinks into separate package objects that a user has to import, to get around any method overloading issues. It would be pretty similar to how they interact with those IOs now actually (like instead of importing `com.spotify.scio.avro._` they'd import `com.spotify.scio.smb.avro._`).

Reply:
Although an uncommon case, but the same method names might still cause problem down the road (say if we add more default args) if one wants to save as Avro + TF records in the same code block. So either the user has to isolate import scopes, or we have to prefix method names to distinguish them.

If the type class approach can parameterize Avro/TF/JSON/etc. and enable a single pair of read/write methods that'd be dope. I wonder if we can also use the type class for test stubbing?

Reply:
Yeah, I just feel that the typeclasses are reinventing the `SortedBucketIO` java APIs a bit... like `SmbIO.forAvro[MyClass](codec, schema, etc.).withKey[String]("keyField")` isn't that far away from the user just using the Java API directly like `AvroSortedBucketIO.write(classOf[String], "keyField", schema)`... especially if the Scala API is just a thin wrapper around the Java one.

Maybe for the first iteration of this we can just have basic sink and source APIs that take `SortedBucketIO.Write` and `SortedBucketIO.Read` params directly, with no implicit conversions for avro/json/TF collections? (and giving up the extra wiring for unit testing.) That way, it's simple, usable and can easily be extended in future Scio versions?

Reply:
👍 yeah all the extra plumbing is really to make `JobTest` work while it won't really so 🤷‍♂

Reply:
cool! I'll fix up the APIs today and add some kind of integration testing example to scio-examples.